### PR TITLE
[perf] feat: Add MFU for Qwen3-VL dense

### PR DESCRIPTION
**Pull request by zhihaofang1017**
### What does this PR do?
Add the _estimate_qwen3_vit_flop and _estimate_qwen3_vl_flops function to calculate the FLOPs of Qwen3-VL dense models. Update the test cases to verify the calculation accuracy of Qwen3-VL models.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test
The following is the output result of running the test file.
image

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/utils/flops_counter.py [hunk 1]
```diff
@@ -429,10 +522,10 @@ def __init__(self, config: PretrainedConfig):
                 f"zero."
             )
 
-        self.config = getattr(config, "text_config", config)
+        self.config = config
```
Comment (tardis-key) [inline verl/utils/flops_counter.py]: This change might conflict with other MFU configurations. Please check and test it.

File: verl/utils/flops_counter.py [hunk 2]
```diff
@@ -118,6 +118,99 @@ def _estimate_qwen2_flops(config, tokens_sum, batch_seqlens, delta_time):
...
+
+    # attn all_layer & all_token fwd & bwd flops
+    seqlen_square_sum = 0
+    for seqlen in batch_seqlens:
+        seqlen_square_sum += seqlen * seqlen
+    attn_qkv_flops = 12 * seqlen_square_sum * head_dim * num_attention_heads * num_hidden_layers
+
+    # vit flops
+    images_seqlens = kargs.get("images_seqlens", None)
+    if images_seqlens is not None:
```
Comment (tardis-key) [inline verl/utils/flops_counter.py:154]: Validating 'images_seqlens'  here allows us to remove the extra branch in fsdp_workers

File: verl/workers/fsdp_workers.py [hunk 3]
```diff
@@ -919,7 +919,15 @@ def update_actor(self, data: DataProto):
                 metrics = self.actor.update_policy(data=data)
             delta_time = timer.last
             global_num_tokens = data.meta_info["global_token_num"]
-            estimated_flops, promised_flops = self.flops_counter.estimate_flops(global_num_tokens, delta_time)
+            images_seqlens = data.meta_info["images_seqlens"]
```
Comment (tardis-key) [inline verl/workers/fsdp_workers.py]: this extra branch can be removed

Comment (tardis-key) [inline verl/workers/fsdp_workers.py]: Is this adaptation required in engine_worker?

**Comment by zhihaofang1017** [comment]
Qwen3-VL MFU #4748

File: verl/utils/flops_counter.py (hunk 1 above)
Comment (tardis-key) [inline verl/utils/flops_counter.py:525]: unifying config usage, nice job!

File: verl/workers/fsdp_workers.py (hunk 3 above)
Comment (zhihaofang1017) [reply verl/workers/fsdp_workers.py]: it has been removed

File: verl/utils/flops_counter.py (hunk 2 above)
Comment (zhihaofang1017) [reply verl/utils/flops_counter.py:154]: done

File: verl/utils/dataset/rl_dataset.py [hunk 4]
```diff
@@ -354,6 +354,16 @@ def __getitem__(self, item):
         row_dict["index"] = index
         row_dict["tools_kwargs"] = tools_kwargs
         row_dict["interaction_kwargs"] = interaction_kwargs
+
+        from qwen_vl_utils import process_vision_info
+        images, videos = process_vision_info(row_dict["raw_prompt"], image_patch_size=self.image_patch_size, return_video_metadata=True)
```
Comment (wuxibin89) [inline verl/utils/dataset/rl_dataset.py]: We should not process_vision_info in `RLDataset` since it run in single-controller's dataloader, which will be a bottleneck for large scale multi-modal dataset. https://github.com/volcengine/verl/pull/4759

Comment (wuxibin89) [reply verl/utils/dataset/rl_dataset.py]: process_vision_info has been move to agent loop worker: https://github.com/volcengine/verl/blob/main/verl/experimental/agent_loop/agent_loop.py#L236

Comment (zhihaofang1017) [reply verl/utils/dataset/rl_dataset.py]: I have removed process_vision_info in RLDataset

File: verl/trainer/ppo/ray_trainer.py [hunk 5]
```diff
@@ -1415,7 +1415,13 @@ def fit(self):
 
                     # compute global_valid tokens
                     batch.meta_info["global_token_num"] = torch.sum(batch.batch["attention_mask"], dim=-1).tolist()
-
+                    # get images_seqlens  
+                    images_seqlens_all = []
+                    for multi_modal_input in batch.non_tensor_batch["multi_modal_inputs"]:
+                        image_grid_thw = multi_modal_input["image_grid_thw"]
+                        images_seqlens = torch.repeat_interleave(image_grid_thw[:, 1] * image_grid_thw[:, 2], image_grid_thw[:, 0])
```
Comment (wuxibin89) [inline verl/trainer/ppo/ray_trainer.py]: There's two problem here:
1. we should also count `video_grid_thw`
2. `images_seqlens` is pretty big: O(#samples) * O(#frames), meta_info is replicated and dispatched to all workers, we should keep it small. Better calculate `images_seqlens` and `seqlen_square_sum` ahead.

File: verl/workers/fsdp_workers.py (hunk 3 above)
Comment (wuxibin89) [reply verl/workers/fsdp_workers.py]: Don't forget megatron_workers.py and engine_workers.py. BTW, we will deprecate `fsdp_workers.py` and `megatron_workers.py` soon.

File: verl/trainer/ppo/ray_trainer.py (hunk 5 above)
Comment (zhihaofang1017) [reply verl/trainer/ppo/ray_trainer.py]: 1. Regarding to this PR, we can initially support image-based calculations, and then add video support later. After all, most scenarios still involve images.
2. In agent_loop.py, when calculating multi_modal_inputs, I have directly add the logic for calculating images_seqlen.

**Review by wuxibin89 — APPROVED** [review]
